synapse/storage/engines/_base.py

```diff
@@ -124,3 +125,21 @@ def attempt_to_set_isolation_level(
         Note: This has no effect on SQLite3, as transactions are SERIALIZABLE by default.
         """
         ...
+
+    @staticmethod
+    @abc.abstractmethod
+    def executescript(cursor: CursorType, script: str) -> None:
+        """Execute a chunk of SQL containing multiple semicolon-delimited statements.
+
+        This is not provided by DBAPI2, and so needs engine-specific support.
+        """
+        ...
+
+    @classmethod
+    def execute_script_file(cls, cursor: CursorType, filepath: str) -> None:
+        """Execute a file containing multiple semicolon-delimited SQL statements.
+
+        This is not provided by DBAPI2, and so needs engine-specific support.
+        """
+        with open(filepath, "rt") as f:
+            cls.executescript(cursor, f.read())
```

Comment:
This generates the schema but I'm not sure if this is transaction-safe. I now get warnings of the form

```
2022-09-23 10:23:30.344 UTC [70] WARNING:  there is already a transaction in progress
```

in postgres logs, which seem scary,

Reply:
Oh, I think this is explained by https://github.com/matrix-org/synapse/blob/5409c76b0c06e02bcd3ba95cdb6cd24d2d6e885e/synapse/storage/prepare_database.py#L107-L112

Reply:
Though we can opt out of this behaviour in Psycopg with autocommit=True? See https://www.psycopg.org/docs/usage.html#transactions-control